### PR TITLE
Add terraform module for ssh nodes

### PR DESCRIPTION
**Pull request by kshtsk**

### CODE REVIEW

File: bin/lib/common.mjs [hunk 1]
```diff
@@ -22,6 +22,18 @@ export function terraformDir(){
     return dir(join("terraform", "main", main_dir))
 }
 
+export function terraformVar() {
+    if (env.TERRAFORM_VAR_FILE) {
+        return `-var-file=${env.TERRAFORM_VAR_FILE}`
+    }
+    if (env.TERRAFORM_MAIN_DIR === "ssh") {
+        console.log("WARNING: TERRAFORM_MAIN_DIR=ssh is used. " +
```
Comment (moio) [inline bin/lib/common.mjs:30]: How about adding all necessary information to inputs.tf?

Comment (kshtsk) [reply bin/lib/common.mjs:30]: What do you mean? Hardcode the IP address to the inputs?

Comment (moio) [reply bin/lib/common.mjs:30]: Variables can be specified without a value, then users can either type them at terraform apply time, or specify special environment variables, or add more (potentially .gitignored) .tf files to their working directory to specify their values.

https://developer.hashicorp.com/terraform/language/values/variables#default-values
https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_var_name

File: terraform/modules/k3s/main.tf [hunk 2]
```diff
@@ -36,20 +36,36 @@ resource "ssh_sensitive_resource" "first_server_installation" {
       node_cidr_mask_size    = var.node_cidr_mask_size
       datastore_endpoint     = var.datastore_endpoint
     })
-    destination = "/root/install_k3s.sh"
+    destination = "/tmp/rodeo-install_k3s.sh"
```
Comment (moio) [inline terraform/modules/k3s/main.tf:39]:
```suggestion
    destination = "/tmp/st-install_k3s.sh"
```

per our discussion

File: terraform/main/ssh/inputs.tf [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+variable ssh_user {
+    description = "Default ssh user for nodes"
+    type = string
+    default = "root"
+}
+
+variable ssh_identity_file {
+    description = "Default ssh user identity file"
+    type = string
+    default = null
```
Comment (moio) [inline terraform/main/ssh/inputs.tf:68]: If a variable must be set for the Terraform module to work correctly, it's best not specifying any default, so that Terraform will fail with an error when it is not specified.

Please remove defaults from essential variables.

File: terraform/main/ssh/inputs.tf [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  clusters = concat([local.upstream_cluster], local.downstream_clusters, [local.tester_cluster])
+
+  // k3s-specific
+  first_local_kubernetes_api_port = 7445
+  first_local_http_port           = 9080
+  first_local_https_port          = 9443
+
+  ssh_private_key_path            = "~/.ssh/rodeo"   // generate with `ssh-keygen -t ed25519`
```
Comment (moio) [inline terraform/main/ssh/inputs.tf:55]: Please turn ssh_private_key_path and ssh_public_key_path into variables with defaults both here and in AWS.

File: bin/lib/common.mjs [hunk 5]
```diff
@@ -22,6 +22,18 @@ export function terraformDir(){
...
+                    "Nodes' addresses in json format must be provided " +
+		    "as an absolute file path via TERRAFORM_VAR_FILE")
+    }
```
Comment (moio) [inline bin/lib/common.mjs:33]: Please remove this ssh-specific warning.

Overall design principle: input to Terraform and output from Terraform should be passed "as-is", without any transformations, to the extent possible - because Terraform bits from this project might be used in other contexts.

In this case, it is up to Terraform and the modules to return a meaningful error.

File: bin/lib/common.mjs (hunk 1 above)
Comment (moio) [reply bin/lib/common.mjs:30]: I re-thought this and now believe supporting a var file is a good thing in general (eg. specifying a non-default SSH key file in AWS).

Thanks for the discussion.

File: bin/lib/common.mjs [hunk 6]
```diff
@@ -22,6 +22,18 @@ export function terraformDir(){
     return dir(join("terraform", "main", main_dir))
 }
 
+export function terraformVar() {
+    if (env.TERRAFORM_VAR_FILE) {
```
Comment (moio) [inline bin/lib/common.mjs:26]: I know we discussed that TERRAFORM_VAR_FILE is a bit of a long name, as well as TERRAFORM_MAIN_DIR.

I would like to keep TERRAFORM_VAR_FILE as it is and rename TERRAFORM_MAIN_DIR to TERRAFORM_WORK_DIR, changing its meaning from a name to a (relative/absolute) path.

Overall design principle: input to Terraform and output from Terraform should be passed "as-is", without any transformations, to the extent possible - because Terraform bits from this project might be used in other contexts.

As a consequence, we want to re-use Terraform jargon as it is to the extent possible - I looked up in the docs and it seems like "working directory" and "variable definition files" are official terminology:

https://developer.hashicorp.com/terraform/language/files#the-root-module
https://developer.hashicorp.com/terraform/language/values/variables#variable-definitions-tfvars-files


I am willing to sacrifice a succint name in exchange of an unambiguous one, at least for now, for the sake of easy interop with other software yet to be written.




I also hear you about having this as a commandline option for convenience. Feel free to implement that if it helps you as a separate PR, but be aware that in future a rewrite of this Javascript code in Golang could happen, and that would use the Cobra library which implements this feature (automatic equivalence of commandline options and environment variables) for free.

File: terraform/main/ssh/inputs.tf [hunk 7]
```diff
@@ -0,0 +1,81 @@
+locals {
+  project_name = "rodeo"
```
Comment (moio) [inline terraform/main/ssh/inputs.tf:2]: As there is no decision to rename the project yet, please stick to "st" for project names (and please to get rid of the remaining instances of "moio" still floating around).

If/when the project is renamed to something short and evocative (which I believe being a good thing, just not immediately), we will do a search/replace.

File: terraform/examples/ssh-cluster.tfvars.json [hunk 8]
```diff
@@ -0,0 +1 @@
+ssh-cluster.tfvars
```
Comment (moio) [inline terraform/examples/ssh-cluster.tfvars.json:1]: Please use HCL to define this file instead of JSON - reason is (IIUC) one cannot add comments to JSON, and they would be useful for a copypastable example file.

In any case users will be able to use any of the two formats.

File: bin/lib/common.mjs (hunk 6 above)
Comment (moio) [reply bin/lib/common.mjs:26]: Please also reword "test environments" in DEVELOPING.md. As per the above comment, at this point I do not want to introduce a new concept/jargon such as "workspace", "backend", or "environment" abstracting on top of Terraform but rather reuse Terraform existing concepts and expose them as-is.

File: terraform/main/ssh/inputs.tf [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    distro_version = "v1.24.12+k3s1"
+    agent_labels   = [
+      [{ key : "monitoring", value : "true" }]
+    ]
+    agent_taints = [
+      [{ key : "monitoring", value : "true", effect : "NoSchedule" }]
+    ]
+    
+    // ssh-specific
+    local_name = "upstream.local.gd"
```
Comment (moio) [inline terraform/main/ssh/inputs.tf:17]: Do .local.gd names make sense in the ssh context?

File: terraform/modules/k3s/inputs.tf [hunk 10]
```diff
@@ -47,6 +47,11 @@ variable "ssh_private_key_path" {
   type        = string
 }
 
+variable "ssh_user" {
+  description = "User name of the SSH bastion host"
```
Comment (moio) [inline terraform/modules/k3s/inputs.tf:51]: Why does the k3s module need to know about the existence of a bastion host?

File: terraform/modules/k3s/main.tf [hunk 11]
```diff
@@ -85,12 +101,29 @@ resource "ssh_resource" "additional_server_installation" {
...
     permissions = "0700"
   }
 
   commands = [
-    "/root/install_k3s.sh > >(tee install_k3s.log) 2> >(tee install_k3s.err >&2)"
+    "sudo /tmp/rodeo-install_k3s.sh > >(tee install_k3s.log) 2> >(tee install_k3s.err >&2)",
+    "mkdir -p ~/.kube",
+    "sudo cat /etc/rancher/k3s/k3s.yaml > ~/.kube/config"
```
Comment (moio) [inline terraform/modules/k3s/main.tf:111]: Any reasons to leave those instructions here rather than rolling them into the install_k3s.sh script?

File: terraform/modules/ssh_host/inputs.tf [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  default     = "root"
+}
+
+variable "ssh_bastion_host" {
+  description = "Public name of the SSH bastion host. Leave null for publicly accessible instances"
+  default     = null
+}
+
+variable "ssh_tunnels" {
+  description = "Opens SSH tunnels to this host via the bastion"
```
Comment (moio) [inline terraform/modules/ssh_host/inputs.tf:49]: As discussed, please leave bastion host as a second step to be addressed in a separate PR when we will have a need for one.

For now, bastion-less access can be assumed.

File: terraform/modules/ssh_host/mount_ephemeral.sh [hunk 13]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
Comment (moio) [inline terraform/modules/ssh_host/mount_ephemeral.sh:1]: I do not think this script makes sense outside of the AWS context

**Review by moio — COMMENTED** [review]
Hey @kshtsk, thanks for the work so far. I added inline comments for (I hope) all of the things we have been discussing yesterday.

I would like to merge this when it is in a good shape, including commented sections and cleanups (commented/dead code).

Let me know if there are any questions.

File: terraform/modules/k3s/main.tf (hunk 11 above)
Comment (kshtsk) [reply terraform/modules/k3s/main.tf:111]: Good question, in general we have ssh access to the nodes as non-root user with sudo permissions. Thus we have to run `install_k3s` via sudo, correspondingly within the sudo execution the script has no idea which user we are running.
PS: there is another side effect, on non-root users `sudo kubectl` does not usually work, because /usr/local/bin is not registered in sudoers as a secure_path, that's why `sudo kubectl` will fail, and we need the k3s config to be copied to `$HOME/.kube/config`.

File: terraform/modules/k3s/inputs.tf (hunk 10 above)
Comment (kshtsk) [reply terraform/modules/k3s/inputs.tf]: Nice catch, it is just an SSH user, not bastion. I'm just parameterizing the hardcode 'root' value.

File: bin/lib/common.mjs (hunk 6 above)
Comment (kshtsk) [reply bin/lib/common.mjs:26]: @moio I've submitted another PR for this, take a look #6

File: terraform/main/ssh/inputs.tf (hunk 9 above)
Comment (kshtsk) [reply terraform/main/ssh/inputs.tf:17]: That's right, need to drop it.

File: terraform/modules/ssh_host/mount_ephemeral.sh (hunk 13 above)
Comment (kshtsk) [reply terraform/modules/ssh_host/mount_ephemeral.sh]: dropped it

File: charts/wordpress/Chart.yaml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v2
```
Comment (moio) [inline charts/wordpress/Chart.yaml:1]: Can you please move the whole Wordpress installation to a separate PR? Thanks!

File: terraform/modules/k3s/main.tf (hunk 11 above)
Comment (kshtsk) [reply terraform/modules/k3s/main.tf:111]: I just kept the logic from k3s module as it was before. Don't see a big reason to drop right now.

Comment (kshtsk) [reply terraform/modules/k3s/main.tf:111]: https://github.com/moio/scalability-tests/blob/master/terraform/modules/k3s/main.tf#L49-L53
The `/root/wait_for_k8s.sh` is executed remotely on the host and using kubectl to poll the cluster status.

Comment (kshtsk) [reply terraform/modules/k3s/main.tf:111]: As I mentioned before, sudo wait_for_k8s will not work, because by default sudoers does not treat /usr/local/bin as secure so kubectl cannot be found in PATH.

Comment (moio) [reply terraform/modules/k3s/main.tf:111]: IIRC kubectl comes from the k3s installer script. If that is the case, feel free to hardcode /usr/local/bin/kubectl in wait_for_k8s since we already assume this is a k3s installation.

Comment (kshtsk) [reply terraform/modules/k3s/main.tf:109]: updated the code

Comment (kshtsk) [reply terraform/modules/k3s/main.tf]: Right, there were two of them, pushed another update

**Comment by moio** [comment]
Rebased and merged with some follow-up changes.

Thanks for the collaboration!
